src/kernel.cpp

```diff
@@ -303,12 +83,13 @@ bool cvk_kernel::setup_descriptor_sets(
     std::lock_guard<std::mutex> lock(m_lock);
 
     // Allocate descriptor sets
+    auto& descriptor_set_layouts = m_entry_point->descriptor_set_layouts();
     VkDescriptorSetAllocateInfo descriptorSetAllocateInfo = {
         VK_STRUCTURE_TYPE_DESCRIPTOR_SET_ALLOCATE_INFO, nullptr,
-        m_descriptor_pool,
+        m_entry_point->descriptor_pool(),
```

Comment:
The allocation code below would violate Vulkan threading requirements on the descriptor pool when multiple kernels using the same entry point allocate descriptors concurrently. Whatever object owns the descriptor pool should be the one responsible for descriptor allocation if we don't want the kernel to have to request the entry point to lock its descriptor pool.

Having the kernel own the descriptor pool enables applications to create several kernels to deal with the `MAX_INSTANCES` limit or reduce contention on the descriptor pool in multithreaded submission scenarios but this would use more memory. I can't think of a killer argument in favour of one solution or the other so I suggest we do whatever is the simplest.

Reply:
Good point.

Given that `cvk_entry_point` owns the descriptor layouts, it makes sense to me to have it own the descriptor pool and do the allocations/frees. This does require it to mutex around these routines as you point out, but not sure whether that's a performance issue in practice. I'd also be reluctant to tie the lifetime of a descriptor pool to a `cl_kernel` object since some workloads (e.g. MACE) seem to create new kernels for every enqueue, so this would be expensive.

I'll push a change to move allocation/free to `cvk_entry_point` (with locks).

Reply:
I don't think owning the layout is a good argument in favour: there is only one layout for each entry point but we have the choice for descriptor pools. That being said, I'm not worried about performance either and what you're proposing sounds good.